library/src/actions/creditCard/creditCard.ts

```diff
@@ -0,0 +1,143 @@
+
+/**
+ * Credit card issue type.
+ */
+export interface CreditCardIssue<TInput extends string>
+  extends BaseIssue<TInput> {
+  /**
+   * The issue kind.
+   */
+  readonly kind: 'validation';
+  /**
+   * The issue type.
+   */
+  readonly type: 'credit_card';
+  /**
+   * The expected input.
+   */
+  readonly expected: null;
+  /**
+   * The received input.
+   */
+  readonly received: `"${string}"`;
+  /**
+   * The validation function.
+   */
+  readonly requirement: (input: string) => boolean;
+}
+
+/**
+ * Credit card action type.
+ */
+export interface CreditCardAction<
+  TInput extends string,
+  TMessage extends ErrorMessage<CreditCardIssue<TInput>> | undefined,
+> extends BaseValidation<TInput, TInput, CreditCardIssue<TInput>> {
+  /**
+   * The action type.
+   */
+  readonly type: 'credit_card';
+  /**
+   * The expected property.
+   */
+  readonly expects: null;
+  /**
+   * The validation function.
+   */
+  readonly requirement: (input: string) => boolean;
+  /**
+   * The error message.
+   */
+  readonly message: TMessage;
+}
+
+/**
+ * Credit card regex.
+ */
+const CREDIT_CARD_REGEX =
+  /^(?:\d{14,19}|\d{4}(?: \d{3,6}){2,4}|\d{4}(?:-\d{3,6}){2,4})$/u;
+
+/**
+ * Sanitize regex.
+ */
+const SANITIZE_REGEX = /[- ]/gu;
+
+/**
+ * Provider regex list.
+ */
+const PROVIDER_REGEX_LIST = [
+  // American Express
+  /^3[47]\d{13}$/u,
+  // Diners Club
+  /^3(?:0[0-5]|[68]\d)\d{11,13}$/u,
+  // Discover
+  /^6(?:011|5\d{2})\d{12,15}$/u,
+  // JCB
+  /^(?:2131|1800|35\d{3})\d{11}$/u,
+  // Mastercard
+  /^5[1-5]\d{2}|(?:222\d|22[3-9]\d|2[3-6]\d{2}|27[01]\d|2720)\d{12}$/u,
+  // UnionPay
+  /^(?:6[27]\d{14,17}|81\d{14,17})$/u,
+  // Visa
+  /^4\d{12}(?:\d{3,6})?$/u,
+];
+
+/**
+ * Creates a credit card validation action.
+ *
+ * @returns A Credit card action.
+ */
+export function creditCard<TInput extends string>(): CreditCardAction<
+  TInput,
+  undefined
+>;
+
+/**
+ * Creates a credit card validation action.
+ *
+ * @param message The error message.
+ *
+ * @returns A credit card action.
+ */
+export function creditCard<
+  TInput extends string,
+  const TMessage extends ErrorMessage<CreditCardIssue<TInput>> | undefined,
+>(message: TMessage): CreditCardAction<TInput, TMessage>;
+
+export function creditCard(
+  message?: ErrorMessage<CreditCardIssue<string>>
+): CreditCardAction<string, ErrorMessage<CreditCardIssue<string>> | undefined> {
+  return {
+    kind: 'validation',
+    type: 'credit_card',
+    expects: null,
+    async: false,
+    message,
+    requirement(input) {
+      let sanitized: string | undefined;
+      return (CREDIT_CARD_REGEX.test(input) &&
+        // Remove any hyphens and blanks
+        (sanitized = input.replace(SANITIZE_REGEX, '')) &&
+        // Check if it matches a provider
+        PROVIDER_REGEX_LIST.some((regex) => regex.test(sanitized)) &&
+        // Check if passes luhn algorithm
+        _isLuhnAlgo(sanitized)) as boolean;
```

Comment:
If you coerce the string value to a boolean with the [double NOT operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_NOT#double_not_!!) to prevent the `as boolean` casting

```suggestion
        !!(sanitized = input.replace(SANITIZE_REGEX, '')) &&
        // Check if it matches a provider
        PROVIDER_REGEX_LIST.some((regex) => regex.test(sanitized)) &&
        // Check if passes luhn algorithm
        _isLuhnAlgo(sanitized));
```

Reply:
Thank you for your review! I considered that, but this code always returns a boolean. So `as boolean` feels right in this case and adds no runtime code.

Reply:
It actually returns `'' | boolean` if the [input or sanitized is an empty string](https://www.typescriptlang.org/play?target=9#code/MYewdgzgLgBAlmADgV1gXhgIgIYCNgAmApgGaYBQANkbBNmHFHAF5EEwaYXmiSwBu2SsiIcYACjoMmrdhgQooAOgBORRJWzAiAQUqVxAeiWGA5gBoYAcisBKWzABkjmAFlsUABar6BEAFtxBwA+GAAGJQBWch5wCBBqJUoQU3EAbxhBYVEAX1sAbiA)

Reply:
The previous `CREDIT_CARD_REGEX.test(input)` check ensures that `sanitized` can't be an empty string. Therefore, I think it is technically impossible for a string to be returned.